drivers/sensor/tdk/icm45686/icm45686_decoder.c

```diff
@@ -369,6 +383,176 @@ static int icm45686_one_shot_decode(const uint8_t *buffer,
 	}
 }
 
+static q31_t icm45686_fifo_read_temp_from_packet(const uint8_t *pkt)
+{
+	struct icm45686_encoded_fifo_payload *fdata = (struct icm45686_encoded_fifo_payload *)pkt;
+
+	int32_t whole;
+	uint32_t fraction;
+	int64_t intermediate;
+	int8_t shift;
+	int err;
+
+	err = icm45686_get_shift(SENSOR_CHAN_DIE_TEMP, 0, 0, &shift);
+	if (err != 0) {
+		return -1;
+	}
+
+	icm45686_temp_c(fdata->temp, &whole, &fraction);
+
+	intermediate = ((int64_t)whole * INT64_C(1000000) + fraction);
+	if (shift < 0) {
+		intermediate =
+			intermediate * ((int64_t)INT32_MAX + 1) * (1 << -shift) / INT64_C(1000000);
+	} else if (shift > 0) {
+		intermediate =
+			intermediate * ((int64_t)INT32_MAX + 1) / ((1 << shift) * INT64_C(1000000));
+	}
+
+	return CLAMP(intermediate, INT32_MIN, INT32_MAX);
+}
+
+static int icm45686_fifo_read_imu_from_packet(const uint8_t *pkt,
+					      bool is_accel,
+					      uint8_t axis_offset,
+					      q31_t *out)
+{
+	uint32_t unsigned_value;
+	int32_t signed_value;
+	int offset = 1 + (axis_offset * 2)  + (is_accel ? 0 : 6);
+	uint32_t mask = is_accel ? GENMASK(7, 4) : GENMASK(3, 0);
+	uint8_t accel_fs = ICM45686_DT_ACCEL_FS_32;
+	uint8_t gyro_fs = ICM45686_DT_GYRO_FS_4000;
+
+	int32_t whole;
+	int32_t fraction;
+	int64_t intermediate;
+	int8_t shift;
+
+	unsigned_value = (pkt[offset] | (pkt[offset + 1] << 8));
+	if (unsigned_value == FIFO_NO_DATA) {
+		return -ENODATA;
+	}
+
+	unsigned_value = (unsigned_value << 4) |
+			 ((pkt[17 + axis_offset] & mask) >> (is_accel ? 4 : 0));
+	signed_value = sign_extend(unsigned_value, 19);
+
+	if (!is_accel) {
+		icm45686_get_shift(SENSOR_CHAN_GYRO_XYZ, accel_fs, gyro_fs, &shift);
+		icm45686_gyro_rads(gyro_fs, signed_value, true, &whole, &fraction);
+	} else {
+		icm45686_get_shift(SENSOR_CHAN_ACCEL_XYZ, accel_fs, gyro_fs, &shift);
+		icm45686_accel_ms(accel_fs, signed_value, true, &whole, &fraction);
+	}
+
+	intermediate = ((int64_t)whole * INT64_C(1000000) + fraction);
+	if (shift < 0) {
+		intermediate =
+			intermediate * ((int64_t)INT32_MAX + 1) * (1 << -shift) / INT64_C(1000000);
+	} else if (shift > 0) {
+		intermediate =
+			intermediate * ((int64_t)INT32_MAX + 1) / ((1 << shift) * INT64_C(1000000));
+	}
+
+	*out = CLAMP(intermediate, INT32_MIN, INT32_MAX);
+
+	return 0;
+}
+
+static int icm45686_fifo_decode(const uint8_t *buffer,
+				struct sensor_chan_spec chan_spec,
+				uint32_t *fit,
+				uint16_t max_count,
+				void *data_out)
+{
+	struct icm45686_encoded_data *edata = (struct icm45686_encoded_data *)buffer;
+	const uint8_t *frame_begin = (const uint8_t *)&edata->fifo_payload;
+	struct icm45686_encoded_fifo_payload *fdata = &edata->fifo_payload;
+	const uint8_t *buffer_end = (edata->header.fifo_count *
+				     sizeof(struct icm45686_encoded_fifo_payload)) +
+				    frame_begin;
+	int count = 0;
+	int err;
+
+	if ((uintptr_t)buffer_end <= *fit || chan_spec.chan_idx != 0) {
```

Comment:
Nit, we have precedence for this so I'm not going to block on it, but we should avoid using `fit` as a pointer. I think it was brought up in a code review a while back that if you move the buffer it invalidates the `fit`. Better to use an offset.

Reply:
Same as https://github.com/zephyrproject-rtos/zephyr/pull/86263#discussion_r1993433192